### PR TITLE
Add gradients for tf.batchNormalization4d

### PR DESCRIPTION
**Pull request by caisq**
support gradients for x, mean, variance, offset and scale.

A step toward fulfilling: https://github.com/tensorflow/tfjs/issues/20

---
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/959)

### CODE REVIEW

**Comment by caisq** [comment]
cc @easadler

**Comment by dsmilkov** [comment]
:lgtm_strong: Thank you! 2 small comments. Feel free to submit after.

---

Reviewed 1 of 2 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[src/ops/batchnorm.ts, line 18 at r2](https://beta.reviewable.io/reviews/tensorflow/tfjs-core/959#-LAG9nl98vMBr_c7jKYp:-LAG9nl98vMBr_c7jKYq:b-21qqpz) ([raw file](https://github.com/tensorflow/tfjs-core/blob/3355b4fb10c1844fe3b4d709480063d27e7c9591/src/ops/batchnorm.ts#L18)):*
> ```XML
>  */
> 
> import {rsqrt} from '..';
> ```

this was probably generated by the editor. Add explicitly the filename

---

*[src/ops/batchnorm.ts, line 199 at r2](https://beta.reviewable.io/reviews/tensorflow/tfjs-core/959#-LAG9R2Z8nq_5ZviSxtc:-LAG9R2_E9JlN-masya7:b-3vkgxf) ([raw file](https://github.com/tensorflow/tfjs-core/blob/3355b4fb10c1844fe3b4d709480063d27e7c9591/src/ops/batchnorm.ts#L199)):*
> ```XML
> 
>     const der = (dy: Tensor) => {
>       util.assert(
> ```

move these asserts to the top of the batchNorm body (outside of the der() function)

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/tensorflow/tfjs-core/959#-:-LAGA9x272NSZfwChLJk:b-8g1atr)*

**Comment by caisq** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[src/ops/batchnorm.ts, line 18 at r2](https://beta.reviewable.io/reviews/tensorflow/tfjs-core/959#-LAG9nl98vMBr_c7jKYp:-LAGGXBz5V7RNiTjmE08:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/3355b4fb10c1844fe3b4d709480063d27e7c9591/src/ops/batchnorm.ts#L18)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

this was probably generated by the editor. Add explicitly the filename
</blockquote></details>

Done.

---

*[src/ops/batchnorm.ts, line 199 at r2](https://beta.reviewable.io/reviews/tensorflow/tfjs-core/959#-LAG9R2Z8nq_5ZviSxtc:-LAGGXTM3IoHGmWdWrxg:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/3355b4fb10c1844fe3b4d709480063d27e7c9591/src/ops/batchnorm.ts#L199)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

move these asserts to the top of the batchNorm body (outside of the der() function)
</blockquote></details>

Done.

---

*[src/ops/batchnorm_test.ts, line 19 at r2](https://beta.reviewable.io/reviews/tensorflow/tfjs-core/959#-LAGAOi_7KLWGruVEJAo:-LAGGfVr5DhXZs4pjfPB:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/3355b4fb10c1844fe3b4d709480063d27e7c9591/src/ops/batchnorm_test.ts#L19)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

in tests, we've been usually calling tf.method() instead of method(). not feeling too strongly.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/tensorflow/tfjs-core/959)*
